src/main/kotlin/com/cognifide/gradle/aem/common/instance/InstanceIMSClient.kt

```diff
@@ -0,0 +1,125 @@
+package com.cognifide.gradle.aem.common.instance
+
+import com.cognifide.gradle.aem.AemExtension
+import com.cognifide.gradle.aem.common.instance.service.ims.Secret
+import io.jsonwebtoken.Jwts
+import io.jsonwebtoken.SignatureAlgorithm
+import org.bouncycastle.openssl.PEMKeyPair
+import org.bouncycastle.openssl.PEMParser
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
+import org.json.JSONObject
+import java.io.*
+import java.security.KeyFactory
+import java.security.KeyPair
+import java.security.interfaces.RSAPrivateKey
+import java.security.spec.KeySpec
+import java.security.spec.PKCS8EncodedKeySpec
+
+/**
+ * Generates Bearer token based on the credentials fetched from AEMaaCS
+ * by communicating with Adobe Identity Management Services.
+ */
+class InstanceIMSClient(private val aem: AemExtension) {
+
+    private val common = aem.common
+
+    /**
+     * URI pointing to credentials file copied from AEMaaCS console.
+     */
+    val serviceCredentialsUrl = aem.obj.string {
+        aem.prop.string("instanceIMSClient.serviceCredentialsUrl")?.let { set(it) }
+    }
+
+    /**
+     * Directory storing the fetched secret file from AEMaaCS instance.
+     */
+    private val secretDir = aem.obj.dir {
+        convention(aem.obj.buildDir("instance/ims"))
+    }
+
+    private val serviceTokenFile: File get() = serviceCredentialsUrl.orNull?.let {
+        common.fileTransfer.downloadTo(it, secretDir.get().asFile)
+    } ?: throw InstanceException("The secret file doesn't exist")
+
+    private lateinit var secret: Secret
+
+    private val expirationTime = aem.obj.long {
+        convention(86400L)
+        aem.prop.long("instanceIMSClient.expirationTime")?.let { set(it) }
+    }
+
+    /**
+     * This is how Adobe calculates expTime in an example they provided for generating token.
+     */
+    private val expTime get() = System.currentTimeMillis() / 1000 + expirationTime.get()
+
+    @Suppress("TooGenericExceptionCaught")
+    fun generateToken(): String? {
+        if (!serviceCredentialsUrl.orNull.isNullOrBlank()) {
+            try {
+                readProperties()
+                val jwtToken = generateJWTToken()
+                return fetchAccessToken(jwtToken)
+            } catch (e: Exception) {
+                throw InstanceException("Couldn't generate the access token, consider checking the provided secret file", e)
+            }
+        }
+        return null
+    }
+
+    private fun readProperties() {
+        secret = serviceTokenFile.inputStream().use {
```

Comment:
remove field `secret`

`return common.formats.toObjectFromJson(it, Secret::class.java)`